pmemkv/src/main/java/site/ycsb/db/pmemkv/PmemKVClient.java

```diff
@@ -119,31 +121,60 @@ public void init() throws DBException {
         String engineName = props.getProperty(ENGINE_PROPERTY, "cmap");
 
         String path = props.getProperty(PATH_PROPERTY);
-        if (path == null) {
-          throw new DBException(PATH_PROPERTY + " is obligatory to run PmemKV client");
-        }
         String size = props.getProperty(SIZE_PROPERTY);
-        if (size == null) {
-          throw new DBException(SIZE_PROPERTY + " is obligatory to run PmemKV client");
+        String jsonConfigProp = props.getProperty(JSON_CONFIG_PROPERTY);
+        Path jsonConfigFile;
+        String jsonConfigContent = null;
+        if (jsonConfigProp != null && !jsonConfigProp.isEmpty()) {
+          jsonConfigFile = Paths.get(jsonConfigProp);
+          try {
+            if (jsonConfigFile != null) {
+              jsonConfigContent = new String(Files.readAllBytes(jsonConfigFile), UTF_8);
+            }
+          } catch (IOException e) {
+            throw new DBException(JSON_CONFIG_PROPERTY +
+              " contain faulty path or points to improper json config params");
+          }
         }
+
         boolean startError = false;
+        Database.Builder<byte[], Map<String, ByteIterator>> builder;
+
+        /* try to open db first */
         try {
-          /* try to open db first */
-          db = new Database.Builder<byte[], Map<String, ByteIterator>>(engineName)
-              .setSize(Long.parseLong(size))
-              .setPath(path)
+          builder = new Database.Builder<byte[], Map<String, ByteIterator>>(engineName);
+          if (jsonConfigContent != null) {
+            builder = builder.fromJson(jsonConfigContent);
+          }
+          if (path != null) {
+            builder = builder.setPath(path);
+          }
+          if (size != null) {
+            /* It's set just in case, it shouldn't be required to open,
+            * but it may be required if someone set up e.g. 'create_if_missing' flag in json file. */
+            builder = builder.setSize(Long.parseLong(size));
+          }
+          db = builder
               .setKeyConverter(new ByteConverter())
               .setValueConverter(new MapToByteBufferConverter())
               .build();
         } catch (DatabaseException e) {
           startError = true;
         }
         if (startError) {
+          /* or create it, if it doesn't exist */
           try {
-            /* or create it, if it doesn't exist */
-            db = new Database.Builder<byte[], Map<String, ByteIterator>>(engineName)
-                .setSize(Long.parseLong(size))
-                .setPath(path)
+            builder = new Database.Builder<byte[], Map<String, ByteIterator>>(engineName);
+            if (jsonConfigContent != null) {
+              builder = builder.fromJson(jsonConfigContent);
+            }
+            if (size != null) {
+              builder = builder.setSize(Long.parseLong(size));
+            }
+            if (path != null) {
+              builder = builder.setPath(path);
+            }
```

Comment:
I've added here a comment, but it wasn't published... The point is to extract this 10 lines as separate function due to the same behavior as above. The only thing different is just comment and order.

Reply:
I'm not sure if it's worth it - we plan to update this (as discussed with Igor) to always read config from json, which will have a default `create_if_missing`, so the second `open` will be removed then.

Reply:
ok, so LG